common/version/semver.go

```diff
@@ -0,0 +1,147 @@
+package version
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// Semver represents a semantic version.
+type Semver struct {
+	major  int
+	minor  int
+	patch  int
+	errata string
+}
+
+// NewSemver creates a new Semver instance.
+func NewSemver(major int, minor int, patch int, errata string) (*Semver, error) {
+	if major < 0 {
+		return nil, fmt.Errorf("major version must be non-negative")
+	}
+	if minor < 0 {
+		return nil, fmt.Errorf("minor version must be non-negative")
+	}
+	if patch < 0 {
+		return nil, fmt.Errorf("patch version must be non-negative")
+	}
+
+	return &Semver{
+		major:  major,
+		minor:  minor,
+		patch:  patch,
+		errata: errata,
+	}, nil
+}
+
+// Parses a semantic version string and returns a Semver instance.
+func SemverFromString(versionStr string) (*Semver, error) {
+	var major int
+	var minor int
+	var patch int
+	var errata string
+
+	if strings.Contains(versionStr, "-") {
+		// Try with errata
+		n, err := fmt.Sscanf(versionStr, "%d.%d.%d-%s", &major, &minor, &patch, &errata)
+		if err != nil {
+			return nil, fmt.Errorf("invalid version format: %w", err)
+		}
+		if n != 4 {
+			return nil, fmt.Errorf("invalid version format")
+		}
+	} else {
+		var extra string
+		n, err := fmt.Sscanf(versionStr, "%d.%d.%d%s", &major, &minor, &patch, &extra)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return nil, fmt.Errorf("invalid version format: %w", err)
+		}
+		if n != 3 {
```

Comment:
Why is this 3, instead of 4?

Reply:
Because `extra` is just to catch if sscanf finds extra characters at the end. Without it, extra characters get ignored.